### PR TITLE
Stop using the "name" argument in the "install-plugin" CLI command

### PR DESCRIPTION
**Pull request by jglick**
Cleans up a longstanding UX regression from https://github.com/jenkinsci/jenkins/pull/2795: there is no good reason to require the `-name` option to the `install-plugin` CLI command. At best it is a nuisance, and at worst it could mean data corruption from typos. So instead of

```sh
java -jar …/jenkins-cli.jar -s http://jenkins/ install-plugin -name some-thing = < target/some-thing.hpi
```

you can just write

```sh
java -jar …/jenkins-cli.jar -s http://jenkins/ install-plugin = < target/some-thing.hpi
```

since the manifest header

```manifest
Short-Name: some-thing
```

already conveys this information.

Ditto for URL-based installation; the former code tried to parse out the short name from the URL path, which was pretty fragile. Now you can just

```sh
java -jar …/jenkins-cli.jar -s http://jenkins/ install-plugin http://download-server/this-naming-is-irrelevant.hpi
```

Also the download is now to a temporary file followed by an atomic move, which should reduce the chance of corrupted files breaking the installation.

Changelog entry:
* The `-name` option to the `install-plugin` command is no longer necessary nor honored.

### CODE REVIEW

**Comment by daniel-beck** [comment]
As a side effect, this will take care of JENKINS-29065 in that "installation" of an invalid file won't work anymore, possibly breaking Jenkins.

**Comment by jglick** [comment]
> "installation" of an invalid file won't work anymore, possibly breaking Jenkins

Confirmed.

**Review by oleg-nenashev — APPROVED** [review]
Looks good to me

**Comment by jglick** [comment]
Test flake:

```
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at hudson.util.RetrierTest.sleepWorksTest(RetrierTest.java:100)
```

File: core/src/main/java/hudson/cli/InstallPluginCommand.java [hunk 1]
```diff
@@ -172,11 +167,20 @@ protected int run() throws Exception {
...
 
-    private static File getTargetFile(String name) {
-        return new File(Jenkins.getActiveInstance().getPluginManager().rootDir,name+".jpi");
+    private static File moveToFinalLocation(File tmpFile) throws Exception {
+        String pluginName;
+        try (JarFile jf = new JarFile(tmpFile)) {
+            pluginName = jf.getManifest().getMainAttributes().getValue("Short-Name");
+        }
+        if (pluginName == null) {
+            throw new IllegalArgumentException("JAR does not look like a plugin");
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/InstallPluginCommand.java]:
```suggestion
            throw new IllegalArgumentException("JAR does not look like a plugin, there is no `Short-Name` manifest attribute");
```

File: core/src/main/java/hudson/cli/InstallPluginCommand.java [hunk 2]
```diff
@@ -172,11 +167,20 @@ protected int run() throws Exception {
...
+    private static File getTmpFile() throws Exception {
+        return File.createTempFile("download", ".jpi.tmp", Jenkins.get().getPluginManager().rootDir);
     }
 
-    private static File getTargetFile(String name) {
-        return new File(Jenkins.getActiveInstance().getPluginManager().rootDir,name+".jpi");
+    private static File moveToFinalLocation(File tmpFile) throws Exception {
+        String pluginName;
+        try (JarFile jf = new JarFile(tmpFile)) {
+            pluginName = jf.getManifest().getMainAttributes().getValue("Short-Name");
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/InstallPluginCommand.java]: Can lead to NPEs for corrupted files, e.g. when there is no manifest
https://docs.oracle.com/javase/7/docs/api/java/util/jar/JarFile.html#getManifest()

**Review by oleg-nenashev — DISMISSED** [review]
Please add checks for Runtime exceptions when reading the manifest (or just for the entire try/catch block)

File: core/src/main/java/hudson/cli/InstallPluginCommand.java (hunk 2 above)
Comment (jglick) [reply core/src/main/java/hudson/cli/InstallPluginCommand.java]: So…the command will fail, as it should. Not an actual problem.

**Review by oleg-nenashev — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
It is rather a bug than a feature, but renaming PR retriggers the build more reliably than close/open

**Pull request merged**